feat: sensor toggles & ui improvements

mozhoku opened the pull request:
## Description

Users can now toggle sensor related things from the UI.

Things can be toggled at will:

- Lidars and lidar visualizations individually
- Traffic camera
- Gnss sensor
- IMU sensor
- Odometry sensor
- Pose sensor

misc changes:
- Added two prefabs for sensor toggles and text in the prefabs folder.
- UI bars now no longer interactable while changing size
- Small change to the UI bars hover color
- The UI tabs will follow each other on size/position change
- Enabled Side panel by default
- Added backgrounds for other objects on the sidebar to prevent text overlap

### Related links

- https://github.com/autowarefoundation/AWSIM-Labs/issues/74

## Tests performed

Tested the binary with Autoware.

https://youtu.be/qH-PHrJ0WRU

## Effects on system behavior

Toggling lidar sensors cause fps drop in the simulator.

## Notes for reviewers

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

lchojnack commented:
@mozhoku FIY 30/31.05 I'm out of the office, then I will review the PR 03.06

xmfcx commented:
Thanks, great work!

Could you:
- make the checkboxes smaller
- make the font smaller
- remove toggle type lines like odometry, pose
- add sensor names to the right of the checkboxes (like lidar top)
- for lidar visibility, add the 👁️ toggle between the checkbox and the sensor name
- width of the sidebar shouldnt change when the size of the window changes
- add vehicle keyboard control as a toggle too
- reset to defaults button should be justified to the left and add label "reset to defaults"

![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/10751153/a9755c2c-6b40-4ab2-b876-01bd46a267de)
![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/10751153/5301e186-ea77-4415-acb5-994ba3c47a8a)

xmfcx commented on Assets/AWSIM/Scripts/UI/UITab.cs:
```diff
@@ -5,55 +5,119 @@ namespace AWSIM.Scripts.UI
+        [SerializeField] public GameObject TabBackground;
+        [SerializeField] private GameObject _uiBar;
+        [SerializeField] private GameObject _barChevron;
+        [SerializeField] private RectTransform _tabTfToFollow;
```
Please rename this file to `UICard.cs`

And the cards should not have reference to each other.

Please check the https://doc.qt.io/qt-6/qvboxlayout.html

The cards should have parent layout container and the container should decide the ordering and positions.

xmfcx replied:
Also check https://doc.qt.io/qt-6/qboxlayout.html

lchojnack commented:
Here is my review regarding the functional part of the feature
- When the Lidar Visualization is off by default (before `Play` is clicked), then the toggle suggests that is on
  ![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/8416175/c2ca48d5-f554-43f5-a350-27057e6a1671)

- I suggest adding a toggle for camera visualization

- I cannot scroll down to for example reset EGO position
  ![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/8416175/f24abdd0-659f-4ca2-99f2-2393c118b047)
 
- As a user, I would expect that when a given sensor is off then the publisher is not visible on the topics list.
  ![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/8416175/4cda19df-353b-4dc0-8993-1707caf32e8c)

- You cannot toggle more than 5 sensors. For example when I've added 7 lidars
  ![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/8416175/24dc7da1-9d1e-4550-a429-9fdda1868f07)

mozhoku commented on Assets/AWSIM/Scripts/UI/Toggle/UISensorToggle.cs:
```diff
@@ -0,0 +1,236 @@
+        private Dictionary<UnityEngine.UI.Toggle, Tuple<Transform, Transform>> _toggleAndLidarParentDictionary;
+        private List<Tuple<Transform, Transform>> _lidarAndParentPairs;
+        private List<PointCloudVisualization> _lidarVisualizations;
+        private List<Transform> _cameraSensors;
```
The script searches the scene for the sensors with their respective tags and adds them to a list. I thought it would be less steps for the users when adding sensor to the scene in the cost of taking control from them. 

I can expose the lists to the editor and add a checkbox for manual list building. Users can toggle whether they want to build the list automatically by with tags or manually thru the editor. :+1:

lchojnack replied:
I think that a checkbox is a good idea. 

I've noticed that the main purpose of  `UI Sensor Toggle` is enabling/disabling unity game objects on the scene. 
For that reason, I'm thinking of reusing the `UI Sensor Toggle` for other purposes for example enabling/disabling the environment sensors as a V2I feature, and ground truth data as EGO ground truth position or NPCs perception result. The User could even toggle another part of the scene like NPC Pedestrians.
I think that in the future new sensors may come like for example radar sensor, and then will be easier to add the sensor to UI.

Moreover, I think that the functionality of searching the sensors on the scene and adding them to the list could be in one script, and the functionality of defining the UI groups and toggle items could be in another script.  

It's just my thoughts.

mozhoku replied:
![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/33583972/2fd29f77-05c0-4625-aaf1-89bb8c9e42f8)

![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/33583972/34d2e50d-5078-4ab2-827f-59c13703a807)

When the manual checkbox is checked, the inspector expands for manual sensor assignment.
After some changes, I've separated the main script into parts and updated the namings. The way UI being built changed a bit but it shouldn't effect the way you want to use it.

mozhoku commented on Assets/AWSIM/Scripts/UI/Toggle/UISensorToggle.cs:
```diff
@@ -0,0 +1,236 @@
+
+namespace AWSIM.Scripts.UI.Toggle
+{
+    public class UISensorToggle : MonoBehaviour
```
I've renamed the class and file name here.

[Assets/AWSIM/Scripts/UI/UISensorInteractionPanel.cs](https://github.com/autowarefoundation/AWSIM-Labs/pull/84/files#diff-4b17f7e45fd11a5baa1a50bf41ebd5168be44f0028ae861fb4a5664019b40237)

mozhoku commented:
Changes after requested reviews:
- Seperated and renamed scripts as suggested.
- Sensor publishers now can't be seen from `ros2 topic echo` when turned off. (sometimes they take long time to turn off (except lidars))
- Sensors can be enabled/disabled properly while keeping their initial inputs.
- Sensor toggles take smaller space in the UI.
- Changed sensor publishers a little bit. Not it is more interectable. (they can be restarted when needed now)
- Changed the way UI being handled. Now it uses proper Unity components for layout management.
- Adjusted UI element sizes etc.
- Made side bar size independent from the simulator window size.
- Side bar can be rescaled by dragging its right edge.
- Added vehicle keyboard control as a toggle to the `Ego Vehicle` panel.

![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/33583972/79470f81-2fb2-4d73-9f28-5a5175ab5dc8)

I'm working on making the sidebar scrollable. It conflicts with resizing for some reason.

xmfcx commented:
It would be better to review once all the changes are finished.

mozhoku commented:
- Added scrollbar for the cards in the side bar.
- Adjusted Graphy visuals.
- Fixed small issues with element scaling, colors, icons.

xmfcx commented:
That is a [scrollview](https://doc.qt.io/qt-5/qml-qtquick-controls2-scrollview.html) without a scrollbar.

mozhoku commented:
@xmfcx I've added the Scroll Bar 
![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/33583972/d3cf5066-fab2-4f76-adac-b1da8f49d38b)

mozhoku commented:
@xmfcx I've moved sidebar to the right as you've suggested.
![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/33583972/296fe01a-f921-45ce-971f-5f08204ba0b3)

mozhoku commented:
@xmfcx I've added margin for the UI cards:
![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/33583972/5f40b640-7386-4364-b20c-0b3b0ad92dad)

mozhoku commented:
@xmfcx 
![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/33583972/fff29151-270d-41dd-9eb4-933452e6b27d)

mozhoku commented:
I've added hotkeys list to the UI. This is how it behaves currently: https://youtu.be/QqhPfXRwe9Y

mozhoku commented:
Current status:
![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/33583972/63d9b306-be01-424a-b266-36d0f164dad8)

Semi transparent when menu is disabled:
![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/33583972/ef8bdd72-6856-4754-9988-7a96b9c775a0)

Aysenayilmaz commented:
While disabling the sensors, the LIDAR topics disappear completely. However, the GNSS, IMU, camera, and pose sensor topics remain in the topic list but do not publish any data. The rest of the changes look good to me .

mozhoku commented:
Update:

- Added toggle to graphics menu for MainCamera: 
![image](https://github.com/autowarefoundation/AWSIM-Labs/assets/33583972/9d2794e7-756e-4960-a43e-14a5e15d04e1)

- Fixed issue with topics still appearing in ros2 topic list after closed from the AWSIM Labs side by forcing garbage collection to run after sensor toggling.
- Changed sensor toggle names on the UI from FrameID to Topic.
- Updated sensor prefabs default tags.
- Separated Graphy from the sideBar.
- Adjusted UI animations lerp speed and scroll speed etc. of the scroll bar.
- Added Hotkeys to UI.

mozhoku commented:
@xmfcx I've added the cursor change on the drag border as you requested:

https://youtu.be/GSb_-H13RQY

xmfcx commented:
Works well for most cases, thanks for your efforts. Could you also update the version to 1.1.1 with this pr? Then a release.

Remaining problems:
- Toggling lidar sensors multiple times cause noticeable lag
- Bird Eye View should start snapped to the vehicle
- Time scale should also be in a card

Could you create issues to solve these?

mozhoku commented:
@xmfcx 
I already have a separate PR for updating the Editor version here https://github.com/autowarefoundation/AWSIM-Labs/pull/101
I'll create issues for the others :+1:
